src/core/thread/discover_scanner.cpp

```diff
@@ -106,12 +110,30 @@ otError DiscoverScanner::Discover(const Mac::ChannelMask &aScanChannels,
     message->SetPanId(aPanId);
     SuccessOrExit(error = Get<Mle>().AppendHeader(*message, Header::kCommandDiscoveryRequest));
 
-    // Append MLE Discovery TLV with a single sub-TLV (MeshCoP Discovery Request).
+    tlv.SetType(Tlv::kDiscovery);
+    SuccessOrExit(error = message->Append(&tlv, sizeof(tlv)));
+
+    startOffset = message->GetLength();
+
+    // Append sub-TLV MeshCoP Discovery Request.
     discoveryRequest.Init();
     discoveryRequest.SetVersion(kThreadVersion);
     discoveryRequest.SetJoiner(aJoiner);
 
-    SuccessOrExit(error = Tlv::AppendTlv(*message, Tlv::kDiscovery, &discoveryRequest, sizeof(discoveryRequest)));
+    SuccessOrExit(error = discoveryRequest.AppendTo(*message));
+
+    if (mAdvDataLength != 0)
+    {
+        // Append sub-TLV MeshCoP Joiner Advertisement.
+        joinerAdvertisement.Init();
+        joinerAdvertisement.SetOui(mOui);
+        joinerAdvertisement.SetAdvData(mAdvData, mAdvDataLength);
+
+        SuccessOrExit(error = joinerAdvertisement.AppendTo(*message));
+    }
+
+    tlv.SetLength(static_cast<uint8_t>(message->GetLength() - startOffset));
+    message->Write(startOffset - sizeof(tlv), sizeof(tlv), &tlv);
```

Comment:
Thanks @lanyuwen. This code works fine. I have a suggestion to simplify it a bit.

Since we create the sub-TLV(s) of `DiscoveryTlv` and therefore their sizes is known beforehand, we can write the Discovery TLV with the correct length in the `message` from the beginning. No need to write it with an unspecified length and then remember the offset in the `message` to re-write the TLV and update its length later. 

```
    discoveryRequest.Init();
    discoveryRequest.SetVersion(kThreadVersion);
    discoveryRequest.SetJoiner(aJoiner);

    if (mAdvDataLength != 0)
    {
        joinerAdvertisement.Init();
        joinerAdvertisement.SetOui(mOui);
        joinerAdvertisement.SetAdvData(mAdvData, mAdvDataLength);
    }

    // Append Discovery TLV with one or two sub-TLVs.
    tlv.SetType(Tlv::kDiscovery);
    tlv.SetLength(static_cast<uint8_t>(discoveryRequest.GetSize() + 
                 ((mAdvDataLength != 0) ? joinerAdvertisement.GetSize() : 0)));

    SuccessOrExit(error = message->Append(&tlv, sizeof(tlv)));
    SuccessOrExit(error = discoveryRequest.AppendTo(*message));

    if (mAdvDataLength != 0)
    {
        SuccessOrExit(error = joinerAdvertisement.AppendTo(*message));
    }
```

Reply:
Great, this is indeed much more simplified. Thank you Abtin! :)

Reply:
Just updated as @abtink  suggested.